src/data/attribute.rs

```diff
@@ -1,137 +1,191 @@
-use crate::data::PropertyPath;
+use crate::data::attribute::errors::PropError;
+use crate::data::{PropertyError, PropertyPath};
 use chrono::{DateTime, FixedOffset};
 use log::{debug, error, warn};
 use protobuf::well_known_types::Struct;
 use serde_json::Value;
 
 pub const KUADRANT_NAMESPACE: &str = "kuadrant";
 
+pub(super) mod errors {
+    use std::error::Error;
+    use std::fmt::{Debug, Display, Formatter};
+
+    #[derive(Debug, PartialEq)]
+    pub enum PropertyError {
+        Get(PropError),
+        Parse(PropError),
+    }
+
+    impl Error for PropertyError {
+        fn source(&self) -> Option<&(dyn Error + 'static)> {
+            match self {
+                PropertyError::Get(err) => Some(err),
+                PropertyError::Parse(err) => Some(err),
+            }
+        }
+    }
+
+    impl Display for PropertyError {
+        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+            match self {
+                PropertyError::Get(e) => {
+                    write!(f, "PropertyError::Get {{ {:?} }}", e)
+                }
+                PropertyError::Parse(e) => {
+                    write!(f, "PropertyError::Parse {{ {:?} }}", e)
+                }
+            }
+        }
+    }
+
+    #[derive(Debug, PartialEq)]
+    pub struct PropError {
+        message: String,
+    }
+
+    impl Display for PropError {
+        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+            write!(f, "PropError {{ message: {:?} }}", self.message)
+        }
+    }
+
+    impl Error for PropError {}
+
+    impl PropError {
+        pub fn new(message: String) -> PropError {
+            PropError { message }
+        }
+    }
+}
+
 pub trait AttributeValue {
-    fn parse(raw_attribute: Vec<u8>) -> Result<Self, String>
+    fn parse(raw_attribute: Vec<u8>) -> Result<Self, PropError>
     where
         Self: Sized;
 }
 
 impl AttributeValue for String {
-    fn parse(raw_attribute: Vec<u8>) -> Result<Self, String> {
+    fn parse(raw_attribute: Vec<u8>) -> Result<Self, PropError> {
         String::from_utf8(raw_attribute).map_err(|err| {
-            format!(
+            PropError::new(format!(
                 "parse: failed to parse selector String value, error: {}",
                 err
-            )
+            ))
         })
     }
 }
 
 impl AttributeValue for i64 {
-    fn parse(raw_attribute: Vec<u8>) -> Result<Self, String> {
-        if raw_attribute.len() != 8 {
-            return Err(format!(
-                "parse: Int value expected to be 8 bytes, but got {}",
-                raw_attribute.len()
-            ));
+    fn parse(raw_attribute: Vec<u8>) -> Result<Self, PropError> {
+        let bytes: Result<[u8; 8], _> = raw_attribute[..8].try_into();
```

Comment:
what if `raw_attribute` has less than 8 bytes?

Reply:
On the line below we check that the conversion is `ok()` and `== 8`: the logic here has been inverted so that the default base case is to error and in the success case returns the value

Reply:
Yeah, no... that's wrong. @eguzki is right, you need to check that it's 8 bytes, _before_ doing the `raw_attribyte[..8]` on line 81, cause [otherwise that'd panic](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=1415c108691a58d41aee7953a9a71f11), as it was doing before. The `Result` is from the `.try_into()`, not the getting the slice into the vector.

Reply:
... or you can just:

`let bytes: Result<[u8; 8], _> = raw_attribute.try_into()`, as that would _only_ work for an 8 byte vector.

Reply:
Ah yeah you're right, I'll address